File: library/src/main/java/ke/tang/ruler/DrawableMarker.java
package ke.tang.ruler;

import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.os.Parcel;
import android.support.annotation.DrawableRes;
import android.view.View;

import java.util.UUID;
import java.util.WeakHashMap;

public class DrawableMarker extends ClickableMarker {
    private static WeakHashMap<String, Drawable> sDrawableCaches = new WeakHashMap<>();
    private transient Drawable mDrawable;
    @DrawableRes
    private int mDrawableRes;
    private int mValue;
    private String mDrawableCacheKey = UUID.randomUUID().toString();

    public DrawableMarker(@DrawableRes int res, int value) {
        mDrawableRes = res;
        mValue = value;
    }

    public DrawableMarker(Drawable drawable, int value) {
        mDrawable = drawable;
        mValue = value;
    }

    @Override
    public void onDraw(Canvas canvas) {
        if (null != mDrawable) {
            mDrawable.setBounds(0, 0, mDrawable.getIntrinsicWidth(), mDrawable.getIntrinsicHeight());
            mDrawable.draw(canvas);
        }
    }

    @Override
    public void onAttach(View v) {
        super.onAttach(v);
        if (0 != mDrawableRes) {
            mDrawable = v.getContext().getResources().getDrawable(mDrawableRes);
        }
    }

    @Override
    public void getBounds(Rect rect) {
        if (null != mDrawable) {
            rect.set(0, 0, mDrawable.getIntrinsicWidth(), mDrawable.getIntrinsicHeight());
        }
    }

    @Override
    public int value() {
        return mValue;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(this.mDrawableRes);
        dest.writeInt(this.mValue);
        dest.writeString(this.mDrawableCacheKey);
        if (0 == mDrawableRes && null != mDrawable) {
            sDrawableCaches.put(mDrawableCacheKey, mDrawable);
        }
    }

    protected DrawableMarker(Parcel in) {
        this.mDrawableRes = in.readInt();
        this.mValue = in.readInt();
        this.mDrawableCacheKey = in.readString();
        if (0 == mDrawableRes) {
            mDrawable = sDrawableCaches.get(mDrawableCacheKey);
            sDrawableCaches.remove(mDrawableCacheKey);
        }
    }

    public static final Creator<DrawableMarker> CREATOR = new Creator<DrawableMarker>() {
        @Override
        public DrawableMarker createFromParcel(Parcel source) {
            return new DrawableMarker(source);
        }

        @Override
        public DrawableMarker[] newArray(int size) {
            return new DrawableMarker[size];
        }
    };
}

File: library/src/main/java/ke/tang/ruler/DrawableMarker.kt
package ke.tang.ruler

import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.os.Parcel
import android.os.Parcelable
import android.view.View
import androidx.annotation.DrawableRes
import java.util.*

class DrawableMarker : ClickableMarker {
    @Transient
    private var drawable: Drawable? = null

    @DrawableRes
    private val drawableRes: Int
    private val value: Int

    constructor(@DrawableRes drawableRes: Int, value: Int, onMarkerClickListener: OnMarkerClickListener? = null) : super(onMarkerClickListener) {
        this.drawableRes = drawableRes
        this.value = value
    }


    constructor(parcel: Parcel) : super(parcel) {
        drawableRes = parcel.readInt()
        value = parcel.readInt()
    }

    override fun onDraw(canvas: Canvas) {
        drawable?.also {
            it.setBounds(0, 0, it.intrinsicWidth, it.intrinsicHeight)
            it.draw(canvas)
        }
    }

    override fun getBounds(rect: Rect) {
        rect.set(0, 0, drawable?.intrinsicWidth ?: 0, drawable?.intrinsicHeight ?: 0)
    }

    override fun value() = value

    override fun onAttach(v: View) {
        if (drawableRes != View.NO_ID) {
            drawable = v.resources.getDrawable(drawableRes)
            drawableCaches.entries.removeAll { it.value == cacheKey }
            drawableCaches[drawable] = cacheKey
        }
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        super.writeToParcel(dest, flags)
        dest.writeInt(drawableRes)
        dest.writeInt(value)
    }

    override fun describeContents(): Int = 0

    companion object {
        val drawableCaches = WeakHashMap<Drawable?, String?>()

        @JvmStatic
        val CREATOR = object : Parcelable.Creator<DrawableMarker> {
            override fun createFromParcel(parcel: Parcel): DrawableMarker {
                return DrawableMarker(parcel)
            }

            override fun newArray(size: Int): Array<DrawableMarker?> {
                return arrayOfNulls(size)
            }
        }

    }
}